---
title: "SQUAREM Gene Level Poisson Model"
author: "Kushal Kumar Dey"
date: "Friday, March 06, 2015"
output: html_document
---

In this script, I consider the gene leve Poisson model. In the previous script, I considered the Poisson reads model which is an analog of Matt Taddy's version of topicmodels. In this case, we consider the version corresponding to the LDa method of Blei and Lafferty. The modification we incorporate i that our data is Poisson and there are random effects (both systematic/batch effects and overdispersion terms). We first present the simulation set up code that builds the counts matrix. 

```{r,echo=TRUE,eval=TRUE}

##################   Simulation set up (Gene Level Poisson) #################

K=4;
G=100;
N=500;

alpha_true=matrix(rnorm((K)*G,1,1),nrow=(K)); ### the matrix of fixed effects

Label.Batch=c(rep(1,N/2),rep(2,N/2)); ##  the batch labels

B=max(Label.Batch);

sigmab_true=1;

beta_true=matrix(0,B,G);       ###  the matrix of the random effect

for(g in 1:G)
{
  beta_true[,g]=rnorm(B,mean=0,sd=sigmab_true);
}

library(gtools)
T=10;
omega_true=matrix(rbind(rdirichlet(T*10,c(3,4,2,6)),rdirichlet(T*10,c(1,4,6,3)),
			rdirichlet(T*10,c(4,1,2,2)),rdirichlet(T*10,c(2,6,3,2)),
			rdirichlet(T*10,c(3,3,5,4))), nrow=N);

###  generating the table 



over_dis=0.3;

noise_true=matrix(0,N,G);

for(n in 1:N)
{
	noise_true[n,]=over_dis*rnorm(G,0,1);
}


read_counts=matrix(0,N,G);
indicator=matrix(0,N,G);

for(n in 1:N)
{
	for(g in 1:G)
	{
		index=sample(1:K,1,omega_true[n,],replace=T);
		mean=exp(alpha_true[index,g] +beta_true[Label.Batch[n],g]+noise_true[n,g]);
		read_counts[n,g]=rpois(1,mean);
		indicator[n,g]=index;
	}
}

k=K
windows()
barplot(t(omega_true),col=2:(k+1),axisnames=F,space=0,border=NA,main="",las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)
title(main=paste("Structure Plot of true topic proportions,k=",K))


#########  Topic model fit   ####################

############  usual topic model  fit  ##########################

	library(maptpx)

	Topic_Clus=topics(read_counts,K,kill=0,tol=0.01);
	docweights_topics=Topic_Clus$omega;
	library(permute);
	library("BioPhysConnectoR");
	perm_set=rbind(1:K,allPerms(1:K));
	diff=array(0,dim(perm_set)[1]);
	for (p in 1:dim(perm_set)[1])
	{
		temp=docweights_topics[,perm_set[p,]];
		diff[p]=fnorm(temp,omega_true);
	}

	p_star=which(diff==min(diff));
	docweights_topics=docweights_topics[,perm_set[p_star,]];

	windows()
	par(mar=c(8,5.25,2.5,2.5))

	# - get rid of space space between leftmost bar and y axis
	
	k=K;

	
	windows()
	barplot(t(docweights_topics),col=2:(k+1),axisnames=F,space=0,border=NA,main="",las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)
  title(main=paste("Structure Plot of topicmodel topic proportions,k=",K))

```

Now we provide the tranformation and reverse transformation functions that convert the simples vector $\omega$ to $\mathbb{R}^{k}$ and back. 

```{r,echo=TRUE,eval=TRUE}
reverse_transform=function(x) 
  {
  	out=array(0,K-1);
		for(i in 1:(K-1))
		{
			out[i]=log((x[i]+10^(-5))/(x[K]+10^(-5)));
		}
		return(out)
	}

# Data
  

	# Log-Likelihood
	loglik_norm <- function(u,y,x) sum((y - x %*% u)^2)

	# Transform the parameters: we just have
	# to find a bijection between R^3 
	# and {(a,b,c,d) \in [0,1]^4 : a+b+c+d=1}.

  transform <- function(v) 
  {
  	# Ensure they in [0,1]
  	temp =c(exp(v),1);
  	out=temp/(1+sum(exp(v)));
  	return(out)
	}

```


Now as in the Beta methylation assay, we provide an initial preprocessing algorithm that gives us some preprocessing estimates of the topic proportions $\omega$ and effect sizes $\alpha$. The main modification to the vase of Beta methylation case we considered is that there will be random effects and overdispersion effects involved in this modeling.

```{r,echo=TRUE,eval=TRUE}
counts=read_counts;

scale=1; K=4; N=500; G=100;

####  The starting value of omega (topic prop weights)  ####################

# Use a preset seed so the example is reproducable.
require("setRNG")
old.seed <- setRNG(list(kind="Mersenne-Twister", normal.kind="Inversion",
seed=54321))

omega_preprocess=matrix(rdirichlet(N,c(scale/K,scale/K,scale/K,scale/K)), nrow=N);

log_counts=log(counts+1e-07);
y=matrix(log_counts,1,N*G);

omega0=omega_preprocess
alpha0=matrix(rnorm((K)*G,1,1),nrow=(K));
param0=c(matrix(omega0,1,N*K),matrix(alpha0,1,K*G));


library(SQUAREM)

counts_batch_process.em <- function(param_vec_batch,y_batch)
{
  Nb=length(y_batch)/G;
	log_counts_batch=matrix(y_batch,Nb,G);
	omega0_batch=matrix(param_vec_batch[1:(Nb*K)],Nb,K);
	alpha0_batch=matrix(param_vec_batch[-(1:(Nb*K))],K,G);

	temp2=t(omega0_batch)%*%log_counts_batch;
	# temp2=t(omega0_batch)%*%pnorm_meth;
	temp1=solve(t(omega0_batch)%*%omega0_batch);
	H = temp1%*%temp2;

###  Estimation of the matrix W (or omega) 

	omega_batch=matrix(0,Nb,K);
	for(n in 1:Nb)
	{
		omega_vec=omega0_batch[n,];
		counts_vec=log_counts_batch[n,];
		res=optim(reverse_transform(omega_vec), function(v) loglik_norm(transform(v),counts_vec,t(H)) );
		omega_batch[n,]=transform(res$par);
	}
	param_vec_omega=matrix(omega_batch,1,Nb*K);
	param_vec_alpha=matrix(H,1,K*G);
	param_new_vec=c(param_vec_omega,param_vec_alpha);
	param_vec=param_new_vec
	return(param_new_vec)
}



omega0=omega_preprocess
alpha0=matrix(rnorm((K)*G,1,1),nrow=(K));
param0=c(matrix(omega0,1,N*K),matrix(alpha0,1,K*G));



counts_process.em <- function(param_vec,y)
{
	log_counts=matrix(y,N,G);
	omega0=matrix(param_vec[1:(N*K)],N,K);
	alpha0=matrix(param_vec[-(1:(N*K))],K,G);
	mean_effect=array(0,c(B,K,G));

	omega=matrix(0,N,K);
	alpha=matrix(0,K,G);

	for(b in 1:B)
	{
		log_counts_batch=log_counts[which(Label.Batch==b),];
		omega0_batch=omega0[which(Label.Batch==b),];
		Nb=length(which(Label.Batch==b));
		y_batch=matrix(log_counts_batch,1,Nb*G)
		alpha0_batch=matrix(rnorm((K)*G,1,1),nrow=(K));
		param0_batch=c(matrix(omega0_batch,1,Nb*K),matrix(alpha0_batch,1,K*G));

		options(warn=-1)
		system.time(res <- squarem(p=param0_batch,y_batch=y_batch, 
		fixptfn=counts_batch_process.em, control=list(maxiter = 30, trace = FALSE)));

		omega_batch=matrix(res$par[1:(Nb*K)],Nb,K);
		alpha_batch=matrix(res$par[-(1:(Nb*K))],K,G);
		
		docweights=omega_batch;
		library(permute);
		library("BioPhysConnectoR");
		perm_set=rbind(1:K,allPerms(1:K));
		diff=array(0,dim(perm_set)[1]);
		for (p in 1:dim(perm_set)[1])
		{
			temp=docweights[,perm_set[p,]];
			diff[p]=fnorm(temp,omega_true[which(Label.Batch==b),]);
		}

		p_star=which(diff==min(diff));
		docweights=docweights[,perm_set[p_star,]];

		omega_batch=docweights;

		alpha_batch=alpha_batch[perm_set[p_star,],];

		omega[which(Label.Batch==b),]=omega_batch;
		mean_effect[b,,]=alpha_batch;
	}

	for(k in 1:K)
	{
		for(g in 1:G)
		{
			alpha[k,g]=mean(mean_effect[,k,g]);
		}
	}

	param_vec_omega=matrix(omega,1,N*K);
	param_vec_alpha=matrix(alpha,1,K*G);
	param_new_vec=c(param_vec_omega,param_vec_alpha);
	param_vec=param_new_vec
	return(param_new_vec)
}

options(warn=-1)
system.time(res <- squarem(p=param0,y=y, fixptfn=counts_process.em, 
				control=list(maxiter = 30, trace = FALSE)));


omega_initial=matrix(res$par[1:(N*K)],N,K);
alpha_initial=matrix(res$par[-(1:(N*K))],K,G)

docweights=omega_initial;
library(permute);
library("BioPhysConnectoR");
perm_set=rbind(1:K,allPerms(1:K));
diff=array(0,dim(perm_set)[1]);
for (p in 1:dim(perm_set)[1])
{
	temp=docweights[,perm_set[p,]];
	diff[p]=fnorm(temp,omega_true);
}

p_star=which(diff==min(diff));
docweights=docweights[,perm_set[p_star,]];

omega_initial=docweights;
alpha_initial=alpha_initial[perm_set[p_star,],];

windows()
k=K
barplot(t(omega_initial),col=2:(k+1),axisnames=F,space=0,border=NA,main="",las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)
title(main=paste("Structure Plot of Preprocessing topic proportions,k=",K))

######################  Using the alpha log likelihood to modify alpha est  #########################

alpha0=alpha_initial;
omega0=omega_initial; 


```

The final estimation is given by 

```{r,echo=TRUE,eval=TRUE}

##########  SQUAREM  Final Estimation Mechanism  ########################


counts=read_counts
y=matrix(counts,1,N*G);

scale=1; K=4; N=500; G=100;

####  The starting value of omega (topic prop weights)  ####################

# Use a preset seed so the example is reproducable.
require("setRNG")
old.seed <- setRNG(list(kind="Mersenne-Twister", normal.kind="Inversion",
seed=54321))

omega0=omega_initial;
alpha0=alpha_initial;

sigmab_0=0.3;

beta0=matrix(0,B,G);       ###  the matrix of the random effect

for(g in 1:G)
{
  beta0[,g]=rnorm(B,mean=0,sd=sigmab_0);
}


param0=c(matrix(omega0,1,N*K),matrix(alpha0,1,K*G),matrix(beta0,1,B*G));

library(lme4)

noise=noise_true;

counts_final.em <- function(param_vec,y)
{
	counts=matrix(y,N,G);
	omega0=matrix(param_vec[1:(N*K)],N,K);
	param_vec_hyp=param_vec[-(1:(N*K))];
	beta0=matrix(tail(param_vec,B*G),B,G);
	alpha0=matrix(param_vec_hyp[(1:(K*G))],K,G);

	################   Estimation of alpha  ###########################

	alpha=matrix(0,K,G);
	for(k in 1:K)
	{
		for(g in 1:G)
		{	
			temp1=0; temp2=0;
			for(n in 1:N)
			{
				temp1=temp1+omega0[n,k]*counts[n,g];
				temp2=temp2+omega0[n,k]*exp(beta0[Label.Batch[n],g]+noise[n,g]);
			}
			alpha[k,g]=log(temp1/temp2);
		}	
	}

	###############   Estimation of omega  ############################

	omega=matrix(0,N,K);
	Z=array(0,c(N,K,G));

	for(n in 1:N)
	{
		for(g in 1:G)
		{
			for(k in 1:K)
			{
				lambda=beta0[Label.Batch[n],g]+alpha[k,g]+noise[n,g];
				Z[n,k,g]=omega0[n,k]*dpois(counts[n,g],exp(lambda));
			}
			Z[n,,g]=(Z[n,,g]+1e-07)/sum(Z[n,,g]+1e-07)
		}
	}

	for(n in 1:N)
	{
		for(k in 1:K)
		{
			omega[n,k]=sum(Z[n,k,])/G;
		}
	}

	docweights=omega;
	library(permute);
	library("BioPhysConnectoR");
	perm_set=rbind(1:K,allPerms(1:K));
	diff=array(0,dim(perm_set)[1]);
	for (p in 1:dim(perm_set)[1])
	{
		temp=docweights[,perm_set[p,]];
		diff[p]=fnorm(temp,omega_true);
	}

	p_star=which(diff==min(diff));
	omega=docweights[,perm_set[p_star,]];


	###############  Estimation of beta  #################################

	beta=matrix(0,B,G);

	for(b in 1:B)
	{
		for(g in 1:G)
		{
			temp1=0; temp2=0;

			for(n in 1:N)
			{
				if(Label.Batch[n]==b)
				{
					temp1=temp1+sum(omega[n,])*counts[n,g];
					temp2=temp2+(omega[n,]%*%exp(alpha[,g]+noise[n,g]));
				}
			}
			beta[b,g]=log(temp1/temp2);
		}
	}

	param_vec_omega=matrix(omega,1,N*K);
	param_vec_alpha=matrix(alpha,1,K*G);
	param_vec_beta=matrix(beta,1,B*G);
	param_new_vec=c(param_vec_omega,param_vec_alpha,param_vec_beta);
	param_vec=param_new_vec
	return(param_new_vec)
}


options(warn=-1)
system.time(res <- squarem(p=param0,y=y, fixptfn=counts_final.em, control=list(maxiter = 20, trace = FALSE)));

omega_final=matrix(res$par[1:(N*K)],N,K);
alpha_final=matrix(res$par[-(1:(N*K))],K,G)

docweights=omega_final;
library(permute);
library("BioPhysConnectoR");
perm_set=rbind(1:K,allPerms(1:K));
diff=array(0,dim(perm_set)[1]);
for (p in 1:dim(perm_set)[1])
{
	temp=docweights[,perm_set[p,]];
	diff[p]=fnorm(temp,omega_true);
}

p_star=which(diff==min(diff));
docweights=docweights[,perm_set[p_star,]];

omega_final=docweights;
alpha_final=alpha_final[perm_set[p_star,],];

windows()
k=K
barplot(t(omega_final),col=2:(k+1),axisnames=F,space=0,border=NA,main="",las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)
title(main=paste("Structure Plot of Final estd. topic proportions,k=",K))



```

